### PR TITLE
SDAP-457 - Script to purge DOMS execution data

### PR DESCRIPTION
**Pull request by RKuttruff**
# Tool to purge old & incomplete DOMS/CDMS execution data

## Prerequisites

You need to have the required python packages installed & the nexus webservice packages installed. If running in the nexus-webapp-driver Docker image or a SDAP dev environment, you don't need to worry. Otherwise: 

* Run `python setup.py install` in `incubator-sdap-nexus/analysis` 
* Run `pip install -r requirements.txt` in `incubator-sdap-nexus/tools/domspurge`

## Usage

`python purge.py -u USERNAME -p PASSWORD [CASSANDRA ARGS...] (--before DATETIME | --before-months MONTHS | --keep-completed | --all) [--keep-failed] [--dry-run] [-y]`

### Options
- `-u & -p`: Cassandra username and password
- One of:
  - `--before`: Datetime (ie, '2023-04-11T19:50:00') before which all data will be purged. Should be given in UTC and must not include timezone.
  - `--before-months`: Number of months prior to current date before which all data will be purged. 
  - `--keep-completed`: Keep all completed executions. (only purge uncompleted executions)
  - `--all`: Purge ALL data. (drops and re-creates keyspace)
- Cassandra args (optional)
  - `--cassandra`: Cassandra hostname(s) or IP(s). (Default: localhost)
  - `-k / --cassandraKeyspace`: Cassandra keyspace for storing DOMS data. (Default: doms)
  - `--cassandraPort`: Port used to connect to Cassandra. (Default: 9042)
  - `--cassandraProtocolVersion`: The version of the Cassandra protocol the driver should use. (Default: 3)
- Additional args
  - `--keep-failed`: Do not purge uncompleted executions (by default all are purged). Incompatible with `--keep-completed`
  - `--dry-run`: Only print the execution ids to be deleted / DB operations to be performed and exit. Do not actually alter the DB
  - `-y / --yes`: Do not prompt the user for confirmation

## Deployment

You can run the script manually in any environment; however, you may wish to automate the process so you don't have to worry about running manual data purges.

To facilitate this, I added the ability to deploy this script as a `CronJob` when deploying SDAP in k8s via the Helm chart. 

Before doing this, you may need to build it into a Docker image (k8s `CronJob`s are run as `Pod`s). We should probably push the image to `nexusjpl/domspurge:1.0.0` as a good default.

### To build the Docker image

```shell
cd /incubator-sdap-nexus
docker build . -f Dockerfile -t <repo>/domspurge:<tag>
```

### Options

Options for the `CronJob` are available in `Values.yml` under `cronJobs.domsPurge`:

| Key              | Description                                                                                    | Default                      | Notes            |
|------------------|------------------------------------------------------------------------------------------------|------------------------------|------------------|
| enabled          | Enable this CronJob in the deployment                                                          | `true`                       |                  |
| schedule         | Schedule for this Job in Cron syntax                                                           | `"0 0 * * *"`                |                  |
| keepBeforeMonths | Age in months of data to purge. (ie, "2" = all executions older than 2 months will be deleted) | `"2"`                        | Must be a string |
| keepUncompleted  | If true, keep failed executions (currently, this is regardless of age)                         | `false`                      |                  |
| image            | Script Docker image                                                                            | `"nexusjpl/domspurge:1.0.0"` |                  |
| restartPolicy    | Job Pod template RestartPolicy                                                                 | `Never`                      |                  |
| concurrency      | ConcurrencyPolicy as defined in `CronJob` API spec                                             | `Forbid`                     |                  |

### CODE REVIEW

**Comment by skorper** [comment]
This is great! I would add a README here: `tools/domspurge/` and add the body from your PR which is really helpful

File: tools/domspurge/purge.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+from dateutil.relativedelta import relativedelta
+from six.moves import input
+from tqdm import tqdm
+
+logging.getLogger('webservice.NexusHandler').setLevel(logging.CRITICAL)
+
+from webservice.algorithms.doms.DomsInitialization import DomsInitializer
+
+
+logging.basicConfig(level=logging.INFO)
```
Comment (skorper) [inline tools/domspurge/purge.py:36]: Maybe log to stdout as well since this is a script?

Comment (RKuttruff) [reply tools/domspurge/purge.py]: Done

**Comment by RKuttruff** [comment]
> This is great! I would add a README here: `tools/domspurge/` and add the body from your PR which is really helpful

Added the readme

**Comment by RKuttruff** [comment]
I also added a means to deploy the script as a `CronJob` in k8s
- Helm template + associated entries in `values.yml`
- `Dockerfile` and `entrypoint.sh` script to run the tool (k8s `CronJob`s are run as `Pod`s thus requiring an image to run)

**Comment by RKuttruff** [comment]
Tested k8s `CronJob` deployment locally

File: tools/domspurge/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Purge DOMS/CDMS Execution Data
+
+## Prerequisites
+
+_If you are running the `deletebyquery.py` script from within the nexus-webapp-driver Docker image, the following prerequisites are not necessary._
```
Comment (ngachung) [inline tools/domspurge/README.md:5]: Do you mean running the purge.py script here?

Comment (RKuttruff) [reply tools/domspurge/README.md]: @ngachung 
Yes. I copy/pasted that text from deletebyquery's README and forgot to change it. 
Corrected

**Review by ngachung — APPROVED** [review]

**Pull request merged**